### PR TITLE
Added PlayerPreResearchEvent

### PR DESCRIPTION
**Pull request by uiytt**
## Description
This event is called when a player click to unlock a research
This is called called before Research#canUnlock(Player)
The event is not called for the cheat sheet

This even can be very useful, with it, devs can deactivate the default research system and if they want, add their own system. 

## Changes
- Created the Event
- Add the callEvent before canUnlock in ChestSlimefunGuide and BookSlimefunGuide
- Changed the order of canUnlock and hasUnlocked (hasUnlocked now before). After some tests, this doesn't cause any problem, and now, if a player open a guide, receive a research by an admin, and then click on the item to receive the research, there won't be a red message with not enough xp send to the player  

## Checklist
- [X] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them. [X] I'm 90% sure It won't break anything
- [X] I followed the existing code standards and didn't mess up the formatting.
- [X] I did my best to add documentation to any public classes or methods I added.
- [X] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [X] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/ChestSlimefunGuide.java [hunk 1]
```diff
@@ -284,15 +286,20 @@ private void displaySlimefunItem(ChestMenu menu, Category category, Player p, Pl
...
-                        } else {
-                            unlockItem(pl, sfitem, player -> openCategory(profile, category, page));
-                        }
+                    if (profile.hasUnlocked(research)) {
+                        openCategory(profile, category, page);
                     } else {
-                        SlimefunPlugin.getLocalization().sendMessage(pl, "messages.not-enough-xp", true);
+                        PreCanUnlockResearchEvent event = new PreCanUnlockResearchEvent(p,research,profile,sfitem,category);
+                        Bukkit.getPluginManager().callEvent(event);
+                        if(!event.isCancelled()) {
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/ChestSlimefunGuide.java:294]: Formatting
```suggestion
                        PreCanUnlockResearchEvent event = new PreCanUnlockResearchEvent(p, research, profile, sfitem, category);
                        Bukkit.getPluginManager().callEvent(event);
                        if (!event.isCancelled()) {
```

Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/ChestSlimefunGuide.java:294]: There could be an empty line before this statement.
```java
                        PreCanUnlockResearchEvent event = new PreCanUnlockResearchEvent(p, research, profile, sfitem, category);
                        Bukkit.getPluginManager().callEvent(event);
                        
                        if (!event.isCancelled()) {
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/BookSlimefunGuide.java [hunk 2]
```diff
@@ -237,15 +239,20 @@ private void addSlimefunItem(Category category, int page, Player p, PlayerProfil
...
-                    } else {
-                        unlockItem(p, item, pl -> openCategory(profile, category, page));
-                    }
+                if (profile.hasUnlocked(research)) {
+                    openCategory(profile, category, page);
                 } else {
-                    SlimefunPlugin.getLocalization().sendMessage(p, "messages.not-enough-xp", true);
+                    PreCanUnlockResearchEvent event = new PreCanUnlockResearchEvent(p,research,profile,item,category);
+                    Bukkit.getPluginManager().callEvent(event);
+                    if(!event.isCancelled()) {
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/BookSlimefunGuide.java:247]: There could be an empty line before this statement.
```java
                        PreCanUnlockResearchEvent event = new PreCanUnlockResearchEvent(p, research, profile, sfitem, category);
                        Bukkit.getPluginManager().callEvent(event);
                        
                        if (!event.isCancelled()) {
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+import javax.annotation.Nonnull;
+
+public class PreCanUnlockResearchEvent extends Event implements Cancellable {
+
+    private static final HandlerList handlers = new HandlerList();
+    private boolean cancelled = false;
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:20]: No need to initialize this variable. Also, could be moved right below the category variable.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        return this.profile;
+    }
+    @Nonnull
+    public SlimefunItem getSlimefunItem() {
+        return this.slimefunItem;
+    }
+    @Nonnull
+    public Category getCategory() {
+        return this.category;
+    }
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:72]: No need to use `this.` in these methods.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/BookSlimefunGuide.java [hunk 5]
```diff
@@ -237,15 +239,20 @@ private void addSlimefunItem(Category category, int page, Player p, PlayerProfil
...
+                        if (research.canUnlock(p)) {
+                            unlockItem(p, item, pl -> openCategory(profile, category, page));
+                        } else {
+                            SlimefunPlugin.getLocalization().sendMessage(p, "messages.not-enough-xp", true);
+                        }
+                    }
                 }
+
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/BookSlimefunGuide.java:255]: Remove this empty line.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/ChestSlimefunGuide.java [hunk 6]
```diff
@@ -284,15 +286,20 @@ private void displaySlimefunItem(ChestMenu menu, Category category, Player p, Pl
...
+                            if (research.canUnlock(pl)) {
+                                unlockItem(pl, sfitem, player -> openCategory(profile, category, page));
+                            } else {
+                                SlimefunPlugin.getLocalization().sendMessage(pl, "messages.not-enough-xp", true);
+                            }
+                        }
                     }
+
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/ChestSlimefunGuide.java:302]: Remove this empty line.

**Review by poma123 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java (hunk 4 above)
Comment (uiytt) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java]: I really don't think this need javadocs, I can if you really want but  I think this is really quite explanatory by itself.

Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java]: Fair enough.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+/**
+ * This {@link Event} is called whenever a {@link Player} click to unlock a research.
+ * This is called called before {@link Research#canUnlock(Player)}
```
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:20]:
```suggestion
 * This {@link Event} is called whenever a {@link Player} clicks to unlock a research.
 * This is called before {@link Research#canUnlock(Player)}.
```

**Review by svr333 — CHANGES_REQUESTED** [review]
Looks good, however, I am noticing that the code in `ChestSlimefunGuide` from line 289 to 301 and the code in `BookSlimefunGuide` from line 242 to 254 is identical. Not sure if thats okay or if that should become one method that both ends just call.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java (hunk 7 above)
Comment (uiytt) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:20]: Will fix the typo 
As for the same code; yes I will add a method in research like Research#clickInteraction(SlimefunGuideImplementation,Player,Item) 
(Looking for a better name)

Comment (uiytt) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java]: Fixed typo and created : guideClickInteraction method in research

**Comment by uiytt** [comment]
Since it's the first time I'm writing a unit test, I'm not too sure if I did it right

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+public class PreCanUnlockResearchEvent extends Event implements Cancellable {
+
+    private static final HandlerList handlers = new HandlerList();
+
+    private final Player player;
+    private final Research research;
+    private final SlimefunItem slimefunItem;
+
+    private boolean cancelled;
+
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:38]: Remove this empty line please :(((

**Review by poma123 — CHANGES_REQUESTED** [review]
I'm pretty much approving this, however, there are still two little formatting issues. The import locations could be improved later down the line (wink at Cookie), so I didn't touch those.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import me.mrCookieSlime.Slimefun.Objects.Category;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import org.apache.commons.lang.Validate;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:16]: should this not be on top

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import io.github.thebusybiscuit.slimefun4.core.researching.Research;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.BookSlimefunGuide;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.ChestSlimefunGuide;
+import me.mrCookieSlime.Slimefun.Objects.Category;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import org.apache.commons.lang.Validate;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:13]: then this

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
+package io.github.thebusybiscuit.slimefun4.api.events;
+
+import io.github.thebusybiscuit.slimefun4.api.player.PlayerProfile;
+import io.github.thebusybiscuit.slimefun4.core.researching.Research;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.BookSlimefunGuide;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.ChestSlimefunGuide;
+import me.mrCookieSlime.Slimefun.Objects.Category;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import org.apache.commons.lang.Validate;
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:9]: third this

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
+package io.github.thebusybiscuit.slimefun4.api.events;
+
+import io.github.thebusybiscuit.slimefun4.api.player.PlayerProfile;
+import io.github.thebusybiscuit.slimefun4.core.researching.Research;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.BookSlimefunGuide;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.ChestSlimefunGuide;
+import me.mrCookieSlime.Slimefun.Objects.Category;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:8]: this as last

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java (hunk 12 above)
Comment (uiytt) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java]: I think I've done what you wanted (not sure), although I'm not convinced of the utility of sorting imports...

File: src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/researches/TestResearches.java [hunk 13]
```diff
@@ -182,4 +187,29 @@ void testUnlockableResearch() {
...
+        NamespacedKey key = new NamespacedKey(plugin, "simple_research");
+        Research research = new Research(key, 250, "Test", 10);
+        research.register();
+
+        SlimefunGuideImplementation guide = Mockito.mock(SlimefunGuideImplementation.class);
+        Player player = server.addPlayer();
+        PlayerProfile profile = TestUtilities.awaitProfile(player);
+        SlimefunItem sfItem = TestUtilities.mockSlimefunItem(plugin, "RESEARCH_TEST", new CustomItem(Material.TORCH, "&bResearch Test"));
+
+        research.guideClickInteraction(guide,player,profile,sfItem,sfItem.getCategory(),0);
```
Comment (TheBusyBiscuit) [inline src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/researches/TestResearches.java:204]: Formatting.
```suggestion
        research.guideClickInteraction(guide, player, profile, sfItem, sfItem.getCategory(), 0);
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/BookSlimefunGuide.java [hunk 14]
```diff
@@ -215,7 +217,11 @@ private void addSlimefunItem(Category category, int page, Player p, PlayerProfil
 
             ChatComponent component = new ChatComponent(ChatUtils.crop(ChatColor.RED, item.getItemName()) + "\n");
             component.setHoverEvent(new HoverEvent(ChatColor.RESET + item.getItemName(), ChatColor.DARK_RED.toString() + ChatColor.BOLD + SlimefunPlugin.getLocalization().getMessage(p, "guide.locked"), "", ChatColor.GREEN + "> Click to unlock", "", ChatColor.GRAY + "Cost: " + ChatColor.AQUA.toString() + research.getCost() + " Level(s)"));
-            component.setClickEvent(new ClickEvent(key, player -> research(player, profile, item, research, category, page)));
+            component.setClickEvent(new ClickEvent(key, player ->
+                    SlimefunPlugin.runSync(() -> {
+                        research.guideClickInteraction(this, player, profile, item, category, page);
+                    })
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/BookSlimefunGuide.java:223]: Having brackets on a one-liner would be redundant.
Also why don't you just do runSync() inside your interaction method?

```suggestion
                    SlimefunPlugin.runSync(() -> 
                        research.guideClickInteraction(this, player, profile, item, category, page);
                    )
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java [hunk 15]
```diff
@@ -189,6 +195,43 @@ public Research addItems(ItemStack... items) {
...
+     * @param profile The {@link PlayerProfile} of that {@link Player}.
+     * @param sfItem The {@link SlimefunItem} on which the {@link Player} clicked.
+     * @param category The {@link Category} where the {@link Player} was.
+     * @param page The page number of where the {@link Player} was in the {@link Category};
+     *
+     * @see ChestSlimefunGuide
+     * @see BookSlimefunGuide
+     */
+    @ParametersAreNonnullByDefault
+    public void guideClickInteraction(SlimefunGuideImplementation guide, Player player, PlayerProfile profile, SlimefunItem sfItem, Category category, int page) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java:216]: This method name is absolutely horrible...
Please choose a better name, like `onPlayerResearch` or `unlockFromGuide`. But not this :NotLikeThis:

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java [hunk 16]
```diff
@@ -189,6 +195,43 @@ public Research addItems(ItemStack... items) {
...
+     *
+     * @param guide The {@link SlimefunGuideImplementation} used.
+     * @param player The {@link Player} who clicked on the item.
+     * @param profile The {@link PlayerProfile} of that {@link Player}.
+     * @param sfItem The {@link SlimefunItem} on which the {@link Player} clicked.
+     * @param category The {@link Category} where the {@link Player} was.
+     * @param page The page number of where the {@link Player} was in the {@link Category};
+     *
+     * @see ChestSlimefunGuide
+     * @see BookSlimefunGuide
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java:213]: No see tags on methods please.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java [hunk 17]
```diff
@@ -189,6 +195,43 @@ public Research addItems(ItemStack... items) {
         return items;
     }
 
+    /**
+     * Handle what to do when a {@link Player} clicks on an un-researched item in
+     * a {@link SlimefunGuideImplementation}.
+     *
+     * @author TheBusyBiscuit
+     * @author uiytt
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java:203]: No author tags on methods please.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+ * This is called before {@link Research#canUnlock(Player)}.
+ * The event is not called for the cheat sheet.
+ *
+ * @author uiytt
+ *
+ * @see ChestSlimefunGuide
+ * @see BookSlimefunGuide
+ *
+ */
+public class PreCanUnlockResearchEvent extends Event implements Cancellable {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:26]: This class name doesn't really follow any conventions here...
Perhaps just  `PlayerPreResearchEvent`?

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.commons.lang.Validate;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import io.github.thebusybiscuit.slimefun4.core.researching.Research;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.BookSlimefunGuide;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.ChestSlimefunGuide;
+
+/**
+ * This {@link Event} is called whenever a {@link Player} clicks to unlock a research.
+ * This is called before {@link Research#canUnlock(Player)}.
+ * The event is not called for the cheat sheet.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:18]: Event could be a link :eyes:

Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:18]: You could also link to the Cheet Sheet class here.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.apache.commons.lang.Validate;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import io.github.thebusybiscuit.slimefun4.core.researching.Research;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.BookSlimefunGuide;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.ChestSlimefunGuide;
+
+/**
+ * This {@link Event} is called whenever a {@link Player} clicks to unlock a research.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:16]: Research could be a link too

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/researches/TestResearches.java (hunk 13 above)
Comment (uiytt) [reply src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/researches/TestResearches.java:204]: Yeah, I should have seen that

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java (hunk 15 above)
Comment (uiytt) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java:216]: I didn't know how to name it :sob: I asked on the discord several time but nobody answered :( 
Anyway, will change this

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java (hunk 18 above)
Comment (uiytt) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java:26]: Thanks, now I'm terrible at naming things

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java (hunk 16 above)
Comment (uiytt) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/researching/Research.java:213]: Why ? but ok will remove this

**Review by poma123 — APPROVED** [review]
LGTM

**Review by svr333 — DISMISSED** [review]

File: src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/researches/TestResearches.java [hunk 21]
```diff
@@ -182,4 +186,29 @@ void testUnlockableResearch() {
...
+        Player player = server.addPlayer();
+        PlayerProfile profile = TestUtilities.awaitProfile(player);
+        SlimefunItem sfItem = TestUtilities.mockSlimefunItem(plugin, "RESEARCH_TEST", new CustomItem(Material.TORCH, "&bResearch Test"));
+
+        research.unlockFromGuide(guide, player, profile, sfItem, sfItem.getCategory(), 0);
+
+        server.getPluginManager().assertEventFired(PlayerPreResearchEvent.class, event -> {
+            Assertions.assertEquals(player, event.getPlayer());
+            Assertions.assertEquals(research, event.getResearch());
+            Assertions.assertEquals(sfItem,event.getSlimefunItem());
```
Comment (TheBusyBiscuit) [inline src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/researches/TestResearches.java:208]: Missing space
```suggestion
            Assertions.assertEquals(sfItem, event.getSlimefunItem());
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PlayerPreResearchEvent.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+public class PlayerPreResearchEvent extends Event implements Cancellable {
+
+    private static final HandlerList handlers = new HandlerList();
+
+    private final Player player;
+    private final Research research;
+    private final SlimefunItem slimefunItem;
+    private boolean cancelled;
+
+
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PlayerPreResearchEvent.java:37]: One of these empty lines is enough.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Two (albeit nitpicky) formatting issues.
But other than these two, it is good to go.

**Review by TheBusyBiscuit — APPROVED** [review]

**Review by poma123 — APPROVED** [review]

**Pull request merged**